runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1481,49 +1501,70 @@ MM_ObjectAccessBarrier::copyObjectFields(J9VMThread *vmThread, J9Class *objectCl
 		}
 	}
 
-	const UDATA *descriptionPtr = (UDATA *) objectClass->instanceDescription;
-	UDATA descriptionBits = 0;
-	if(((UDATA)descriptionPtr) & 1) {
-		descriptionBits = ((UDATA)descriptionPtr) >> 1;
-	} else {
-		descriptionBits = *descriptionPtr++;
-	}
-
-	UDATA descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
 	UDATA offset = 0;
 	UDATA limit = objectClass->totalInstanceSize;
 	UDATA const referenceSize = J9VMTHREAD_REFERENCE_SIZE(vmThread);
+	bool hasReferences = J9CLASS_HAS_REFERENCES(objectClass);
+	U_32 firstFieldOffset = (U_32) objectClass->backfillOffset;
+
+	if (hasReferences) {
+		const UDATA *descriptionPtr = (UDATA *) objectClass->instanceDescription;
+		UDATA descriptionBits = 0;
+		if(((UDATA)descriptionPtr) & 1) {
+			descriptionBits = ((UDATA)descriptionPtr) >> 1;
+		} else {
+			descriptionBits = *descriptionPtr++;
+		}
 
-	if (isValueType) {
-		U_32 firstFieldOffset = (U_32) objectClass->backfillOffset;
-		if (0 != firstFieldOffset) {
-			/* subtract padding */
-			offset += firstFieldOffset;
-			descriptionBits >>= 1;
-			descriptionIndex -= 1;
+		UDATA descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
+
+		if (isValueType) {
+			if (0 != firstFieldOffset) {
+				/* subtract padding */
+				offset += firstFieldOffset;
+				descriptionBits >>= 1;
+				descriptionIndex -= 1;
+			}
 		}
-	}
 
-	while (offset < limit) {
-		/* Determine if the slot contains an object pointer or not */
-		if(descriptionBits & 1) {
-			J9Object *objectPtr = mixedObjectReadObject(vmThread, srcObject, srcOffset + offset, false);
-			mixedObjectStoreObject(vmThread, destObject, destOffset + offset, objectPtr, false);
-		} else {
+		while (offset < limit) {
+			/* Determine if the slot contains an object pointer or not */
+			if(descriptionBits & 1) {
+				J9Object *objectPtr = mixedObjectReadObject(vmThread, srcObject, srcOffset + offset, false);
+				mixedObjectStoreObject(vmThread, destObject, destOffset + offset, objectPtr, false);
+			} else {
+				UDATA srcAddress = (UDATA)srcObject + srcOffset + offset;
+				UDATA destAddress = (UDATA)destObject + destOffset + offset;
+				if (sizeof(uint32_t) == referenceSize) {
+					*(uint32_t *)destAddress = *(uint32_t *)srcAddress;
+				} else {
+					*(uintptr_t *)destAddress = *(uintptr_t *)srcAddress;
+				}
+			}
+			descriptionBits >>= 1;
+			if(descriptionIndex-- == 0) {
+				descriptionBits = *descriptionPtr++;
+				descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
+			}
+			offset += referenceSize;
+		}
+	} else {
+		/* no instanceDescription bits needed on this path */
+		if (isValueType) {
+			if (0 != firstFieldOffset) {
```

Comment:
There's no need for the zero check here (and elsewhere this pattern appears).

Reply:
We are currently using this to indicate the presence of pre-padding in valuetypes. If this check is skipped then the instance description bits will be misalligned in the flattened representation (since padding is removed in these cases). 

I plan to address this in the next PR.